pkg/authorization/authorizer/authorizer.go

```diff
@@ -0,0 +1,513 @@
+	// handle input of form /api/{version}/* by adjusting special paths
+	if parts[0] == "api" {
+		if len(parts) > 2 {
+			parts = parts[2:]
+		} else {
+			return "", "", "", parts, fmt.Errorf("Unable to determine kind and namespace from url, %v", req.URL)
+		}
+	}
+
+	// handle input of form /{specialVerb}/*
+	if _, ok := specialVerbs[parts[0]]; ok {
+		verb = parts[0]
+		if len(parts) > 1 {
+			parts = parts[1:]
+		} else {
+			return "", "", "", parts, fmt.Errorf("Unable to determine kind and namespace from url, %v", req.URL)
+		}
+	}
+
+	// URL forms: /ns/{namespace}/{kind}/*, where parts are adjusted to be relative to kind
+	if parts[0] == "ns" {
+		if len(parts) < 3 {
+			return "", "", "", parts, fmt.Errorf("ResourceTypeAndNamespace expects a path of form /ns/{namespace}/*")
+		}
+		return verb, parts[1], parts[2], parts[2:], fmt.Errorf("Unable to determine kind and namespace from url, %v", req.URL)
+	}
+
+	// URL forms: /{kind}/*
+	// URL forms: POST /{kind} is a legacy API convention to create in "default" namespace
+	// URL forms: /{kind}/{resourceName} use the "default" namespace if omitted from query param
+	// URL forms: /{kind} assume cross-namespace operation if omitted from query param
+	kind := parts[0]
+	namespace := req.URL.Query().Get("namespace")
+	if len(namespace) == 0 {
+		if len(parts) > 1 || req.Method == "POST" {
+			namespace = kapi.NamespaceDefault
+		} else {
+			namespace = kapi.NamespaceAll
+		}
+	}
+	return verb, kind, namespace, parts, nil
+}
+
+// splitPath returns the segments for a URL path.
+func splitPath(path string) []string {
+	path = strings.Trim(path, "/")
+	if path == "" {
+		return []string{}
+	}
+	return strings.Split(path, "/")
+}
+
+func GetBootstrapPolicy(masterNamespace string) *authorizationapi.Policy {
+	return &authorizationapi.Policy{
+		ObjectMeta: kapi.ObjectMeta{
+			Name:              authorizationapi.PolicyName,
+			Namespace:         masterNamespace,
+			CreationTimestamp: util.Now(),
+		},
+		LastModified: util.Now(),
+		Roles: map[string]authorizationapi.Role{
+			"cluster-admin": {
+				ObjectMeta: kapi.ObjectMeta{
+					Name:      "cluster-admin",
+					Namespace: masterNamespace,
+				},
+				Rules: append(make([]authorizationapi.PolicyRule, 0),
+					authorizationapi.PolicyRule{
+						Verbs:         append(make([]string, 0), "*"),
+						ResourceKinds: append(make([]string, 0), "*"),
+					}),
+			},
+			"admin": {
+				ObjectMeta: kapi.ObjectMeta{
+					Name:      "admin",
+					Namespace: masterNamespace,
+				},
+				Rules: append(make([]authorizationapi.PolicyRule, 0),
+					authorizationapi.PolicyRule{
+						Verbs:         append(make([]string, 0), "*", "-create", "-update", "-delete"),
+						ResourceKinds: append(make([]string, 0), "*"),
+					},
+					authorizationapi.PolicyRule{
+						Verbs:         append(make([]string, 0), "create", "update", "delete"),
+						ResourceKinds: append(make([]string, 0), "*", "-roles", "-policyBindings"),
```

Comment:
add a todo here to think about preventing admin privilege escalation within their namespace by binding themselves to a cluster-admin role, or by creating a local role that allows *,*